application/src/main/java/org/thingsboard/server/service/edge/rpc/EdgeGrpcService.java

```diff
@@ -139,29 +143,32 @@ public void destroy() {
         if (scheduler != null) {
             scheduler.shutdownNow();
         }
+        if (syncExecutorService != null) {
+            syncExecutorService.shutdownNow();
+        }
     }
 
     @Override
     public StreamObserver<RequestMsg> handleMsgs(StreamObserver<ResponseMsg> outputStream) {
-        return new EdgeGrpcSession(ctx, outputStream, this::onEdgeConnect, this::onEdgeDisconnect, mapper).getInputStream();
+        return new EdgeGrpcSession(ctx, outputStream, this::onEdgeConnect, this::onEdgeDisconnect, mapper, syncExecutorService).getInputStream();
     }
 
     @Override
-    public void updateEdge(Edge edge) {
+    public void updateEdge(TenantId tenantId, Edge edge) {
         EdgeGrpcSession session = sessions.get(edge.getId());
         if (session != null && session.isConnected()) {
-            log.debug("[{}] Updating configuration for edge [{}] [{}]", edge.getTenantId(), edge.getName(), edge.getId());
+            log.debug("[{}] Updating configuration for edge [{}] [{}]", tenantId, edge.getName(), edge.getId());
             session.onConfigurationUpdate(edge);
         } else {
-            log.debug("[{}] Session doesn't exist for edge [{}] [{}]", edge.getTenantId(), edge.getName(), edge.getId());
+            log.debug("[{}] Session doesn't exist for edge [{}] [{}]", tenantId, edge.getName(), edge.getId());
         }
     }
 
     @Override
-    public void deleteEdge(EdgeId edgeId) {
+    public void deleteEdge(TenantId tenantId, EdgeId edgeId) {
         EdgeGrpcSession session = sessions.get(edgeId);
         if (session != null && session.isConnected()) {
-            log.info("Closing and removing session for edge [{}]", edgeId);
+            log.info("[{}] Closing and removing session for edge [{}]", tenantId, edgeId);
             session.close();
             sessions.remove(edgeId);
             sessionNewEvents.remove(edgeId);
```

Comment:
Add ConcurrentMap<EdgeId, Lock> and change sessionNewEvents from ConcurrentMap to simple Map. Execure check and update operation inside of the lock to avoid race conditions and updates stuck due to race condition.